File: MaximumDepthofBinaryTree.cpp
#include <iostream>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    int maxDepth(TreeNode *root)
    {
        if (root == nullptr)
            return 0;
        int Lhight = maxDepth(root->left);
        int Rhight = maxDepth(root->right);
        return max(Lhight, Rhight) + 1;
    }
};

int main()
{
    TreeNode *root = new TreeNode(5);
    root->right = new TreeNode(5);
    root->left = new TreeNode(3);
    root->left->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right->left = new TreeNode(1);
    Solution sol;
    cout << sol.maxDepth(root);
    return 0;
}

File: MaximumDepthofBinaryTree.scala
class TreeNode(
    _value: Int = 0,
    _left: TreeNode = null,
    _right: TreeNode = null
) {
  var value: Int = _value
  var left: TreeNode = _left
  var right: TreeNode = _right
}

object Solution {
  def maxDepth(root: TreeNode): Int = {
    if (root == null) 0
    else {
      val Ldepth = maxDepth(root.left)
      val Rdepth = maxDepth(root.right)
      math.max(Ldepth, Rdepth) + 1
    }
  }

  def main(args: Array[String]): Unit = {
    val root = new TreeNode(5)
    root.left = new TreeNode(3)
    root.left.left = new TreeNode(4)
    root.left.right = new TreeNode(3)
    root.left.right.left = new TreeNode(1)
    root.right = new TreeNode(5)

    println(maxDepth(root))
  }
}
